### PR TITLE
Flat tree with binary-content nodes implementation of max-rate registry

### PR DESCRIPTION
**Pull request by dankraw**
Introducing alternative implementation for max-rate registry in which we keep whole max-rate configuration dedicated to a single consumer in just one znode:
`/hermes/consumers-rate/<cluster>/runtime-bin/<consumer-id>/max-rate`
On the other hand consumer writes its current sending rate to a single znode:
`/hermes/consumers-rate/<cluster>/runtime-bin/<consumer-id>/history`

These znodes are encoded in binary format (using [SBE](https://github.com/real-logic/simple-binary-encoding)) and use [numeric ids](https://github.com/allegro/hermes/pull/1078) for subscription identification.

The leader node reads `history` znodes of all consumers on each max-rate computation. After the computations are done, it saves the results to `max-rate` znodes, so consumers can update their sending max-rates.

By using this registry implementation we generate much less znodes in zookeeper and need few zookeeper watches, the structure is flat and should load very fast and take not much space even for thousands of subscriptions controlled by a single consumer node. The number of znodes grows linearly as the number of consumers grow. I haven't implemented any zookeeper watches except a single `NodeCache` per consumer node, so it is immediately notified about max-rate being changed by the leader. Maybe in the future it will be worth implementing a `PathChildrenCache` just for the leader node. 

`HierarchicalCacheMaxRateRegistry` is basically old `MaxRateRegistry` class, it's just renamed, as `MaxRateRegistry` is now an interface, so different implementations are available to choose from. Personally, I think the interface is too wide, but I wanted to leave it as is for now, so we don't change too much in the old max-rate implementation. If we stop supporting it, it will be easy to refactor and for example divide the registry into two parts: one for the leader that makes all the computations (so it reads history znodes of all consumers in its cluster), and second for all the nodes other nodes in which only max-rate znode is loaded and rate history znode is updated.

Note: all classes from `pl.allegro.tech.hermes.consumers.consumer.rate.sbe.stubs` package are generated using SBE build tool via `./gradlew hermes-consumers:generateSbeStubs`.
Schema configuration files that the stubs are generated from are kept in `hermes-consumers/src/main/resources/sbe/` directory.

### CODE REVIEW

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerMaxRateRegistryType.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package pl.allegro.tech.hermes.consumers.consumer.rate.maxrate;
+
+import pl.allegro.tech.hermes.common.exception.InternalProcessingException;
+
+import static java.lang.String.join;
+
+class ConsumerMaxRateRegistryType {
```
Comment (cristaloleg) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerMaxRateRegistryType.java]: enum?

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        // when
+        when(assignedConsumersSupplier.getAllAssignedConsumers()).thenReturn(Collections.emptySet());
+        maxRateRegistry.onBeforeMaxRateCalculation(); // cleanup
+
+        // then
+        wait.untilZookeeperPathNotExists(paths.consumerMaxRatePath(consumerId));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void shouldNotAllowForSavingRateHistoryOfOtherNodes() {
```
Comment (jewertow) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java]: What `other` means in this context? I think it's not clear.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistry.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.slf4j.LoggerFactory.getLogger;
+import static pl.allegro.tech.hermes.common.config.Configs.CONSUMER_WORKLOAD_NODE_ID;
+
+public class FlatBinaryMaxRateRegistry implements MaxRateRegistry, NodeCacheListener {
```
Comment (jewertow) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistry.java]: I think scope of class and constructor could be package-private. (The same for `HierarchicalCacheMaxRateRegistry`)

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java (hunk 2 above)
Comment (dankraw) [reply hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java]: Here it is possible to get maximum rate or rate history only for the current node on which the method is called, it is not possible to get data related to "other" nodes in the cluster. So the leader making the calculation won't be able to fetch data using this method. That's why I've added some preconditions in order not to make a mistake in using it. 

I'm aware the interface is not very clear unfortunately, but I wanted to stick as much as possible to the previous max-registry implementation and not refactor everything.

File: hermes-consumers/src/main/resources/sbe/max-rate.xml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        <group name="Subscriptions" id="1">
+            <field name="Id" id="2" type="int64"/>
+            <field name="MaxRate" id="3" type="double"/>
+        </group>
+    </sbe:message>
+    <sbe:message name="RateHistory" id="2" description="Historic rate of a consumer node">
+        <group name="Subscriptions" id="1" dimensionType="groupSizeEncoding">
+            <field name="Id" id="2" type="int64"/>
+            <group name="Rates" id="3">
+                <field name="Rate" id="4" type="double"/>
```
[the remark this answers is not in the document]
Comment (dankraw) [reply hermes-consumers/src/main/resources/sbe/max-rate.xml]: Umm.. I'm not sure if I understand the question correctly, `group` does not stand for "field grouping" only. Here it means rate-history entry for a single subscriptions can have multiple rates reported. 
It is configurable via [consumer.maxrate.history.size ](https://github.com/allegro/hermes/blob/hermes-1.2.1/hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java#L207)

Currently I'm not planning to add any new fields.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistry.java (hunk 3 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistry.java]: fixed

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/di/ConsumersBinder.java [hunk 5]
```diff
@@ -34,9 +34,11 @@
 import pl.allegro.tech.hermes.consumers.consumer.offset.OffsetQueue;
 import pl.allegro.tech.hermes.consumers.consumer.rate.ConsumerRateLimitSupervisor;
 import pl.allegro.tech.hermes.consumers.consumer.rate.calculator.OutputRateCalculatorFactory;
+import pl.allegro.tech.hermes.consumers.consumer.rate.maxrate.HierarchicalCacheMaxRateRegistry;
```
[the remark this answers is not in the document]
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/di/ConsumersBinder.java]: fixed

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerMaxRateRegistryType.java (hunk 1 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerMaxRateRegistryType.java]: done

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            try {
+                List<String> assignedConsumers = consumersInRegistry(subscriptionName);
+                if (setInitialMaxRates(subscriptionName, assignedConsumers)) {
+                    loadedMaxRates++;
+                }
+            } catch (Exception e) {
+                logger.warn("Exception occurred when initializing cache for subscription {}", subscriptionName, e);
+            }
+        }
+        ;
```
Comment (cristaloleg) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java]: Leftover?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+    @Override
+    public Set<ConsumerRateInfo> ensureCorrectAssignments(SubscriptionName subscriptionName, Set<String> currentConsumers) {
+        Set<ConsumerRateInfo> rateInfos = new HashSet<>();
+        try {
+            cleanupRegistry(subscriptionName, new ArrayList<>(currentConsumers));
+
+            for (String consumerId : currentConsumers) {
+                ConsumerInstance consumerInstance = new ConsumerInstance(consumerId, subscriptionName);
+                RateInfo rateInfo = this.rateInfos.getOrDefault(consumerInstance, RateInfo.empty());
```
Comment (cristaloleg) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java]: `this` isn't needed, isn't it?

Comment (cristaloleg) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java]: ah, I see, there is a `Set rateInfos` above.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+import static java.util.Collections.emptyList;
+
+class HierarchicalCacheMaxRateRegistry implements MaxRateRegistry {
+
+    private static final Logger logger = LoggerFactory.getLogger(HierarchicalCacheMaxRateRegistry.class);
+
+    private static final int LEVEL_SUBSCRIPTION = 0;
+    private static final int LEVEL_CONSUMER = 1;
+    private static final int LEVEL_CONTENT = 2;
```
Comment (cristaloleg) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java]: Looks like a good candidates for enum.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerRateHistoriesDecoder.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                    RateHistoryDecoder.TEMPLATE_ID, header.templateId()));
+        }
+
+        body.wrap(buffer, header.encodedLength(), header.blockLength(), header.version());
+
+        ConsumerRateHistory result = new ConsumerRateHistory();
+        for (RateHistoryDecoder.SubscriptionsDecoder subscriptionDecoder : body.subscriptions()) {
+            long id = subscriptionDecoder.id();
+            List<Double> rates = new ArrayList<>();
+            for (RateHistoryDecoder.SubscriptionsDecoder.RatesDecoder ratesDecoder : subscriptionDecoder.rates()) {
```
Comment (cristaloleg) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerRateHistoriesDecoder.java]: Stream api? to get rid of long `RateHistoryDecoder.SubscriptionsDecoder.RatesDecoder ratesDecoder`. WDYT?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java (hunk 6 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/HierarchicalCacheMaxRateRegistry.java]: dunno, missclick maybe :) fixed

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/MaxRateRegistryFactory.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            throw e;
+        }
+        logger.info("Max rate registry type chosen: {}", type.getConfigValue());
+
+        switch (type) {
+            case HIERARCHICAL:
+                return new HierarchicalCacheMaxRateRegistry(configFactory, curator, objectMapper, zookeeperPaths, pathSerializer, subscriptionCache);
+            case FLAT_BINARY:
+                return new FlatBinaryMaxRateRegistry(configFactory, subscriptionAssignmentCache::getAssignedConsumers,
+                        subscriptionAssignmentCache::getConsumerSubscriptions, curator, zookeeperPaths, subscriptionIds);
```
[the remark this answers is not in the document]
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/MaxRateRegistryFactory.java]: done

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerRateHistoriesDecoder.java (hunk 9 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/ConsumerRateHistoriesDecoder.java]: done

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java (hunk 2 above)
Comment (jewertow) [reply hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/rate/maxrate/FlatBinaryMaxRateRegistryTest.java]: Thanks for the explanation.

**Pull request merged**